lib/ansible/galaxy/collection/concrete_artifact_manager.py

```diff
@@ -66,15 +67,72 @@ class ConcreteArtifactsManager:
         * retrieving the metadata out of the downloaded artifacts
     """
 
-    def __init__(self, b_working_directory, validate_certs=True):
-        # type: (bytes, bool) -> None
+    def __init__(self, b_working_directory, validate_certs=True, keyring=None):
+        # type: (bytes, bool, str) -> None
         """Initialize ConcreteArtifactsManager caches and costraints."""
         self._validate_certs = validate_certs  # type: bool
         self._artifact_cache = {}  # type: Dict[bytes, bytes]
         self._galaxy_artifact_cache = {}  # type: Dict[Union[Candidate, Requirement], bytes]
         self._artifact_meta_cache = {}  # type: Dict[bytes, Dict[str, Optional[Union[str, List[str], Dict[str, str]]]]]
         self._galaxy_collection_cache = {}  # type: Dict[Union[Candidate, Requirement], Tuple[str, str, GalaxyToken]]
+        self._galaxy_collection_origin_cache = {}  # type: Dict[Candidate, Tuple[str, List[Dict[str, str]]]]
         self._b_working_directory = b_working_directory  # type: bytes
+        self._supplemental_signature_cache = {}  # type: Dict[str, str]
+        self._keyring = keyring  # type: str
+
+    @property
+    def keyring(self):
+        return self._keyring
+
+    def get_signature_from_source(self, source: str, display: Display = None) -> str:
```

Comment:
Looks like this only deals with the external resources on the internet. But the artifact manager is primarily expected to work with artifacts on disk.
I would've imagined that it could return signatures of collections that are on disk but not something that is not even downloaded. Maybe it could have a candidate as an input and use a URL out of it + use that candidate as a cache key?

Reply:
This was initially in the gpg file, but you suggested it might belong here. You're totally right that this has nothing to do with artifacts on disk, so I moved it back.